File: OneQuestionPerDay/53/maximum_subarray2.py
from typing import List

from collections import namedtuple

status = namedtuple("Status", ["l_sum", "r_sum", "i_sum", "m_sum"])


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        return self.get(nums, 0, len(nums) - 1).m_sum

    def push_up(self, l: status, r: status) -> status:
        return status(
            max(l.l_sum, l.i_sum + r.l_sum),
            max(r.r_sum, l.r_sum + r.i_sum),
            l.i_sum + r.i_sum,
            max(l.m_sum, l.r_sum + r.l_sum, r.m_sum),
        )

    def get(self, arr: List[int], l: int, r: int) -> status:
        if l == r:
            return status(arr[l], arr[l], arr[l], arr[l])

        m = (l + r) // 2
        l_sub = self.get(arr, l, m)
        r_sub = self.get(arr, m + 1, r)
        return self.push_up(l_sub, r_sub)

File: OneQuestionPerDay/53/maximum_subarray2.go
package main

type status struct {
	lSum int
	rSum int
	iSum int
	mSum int
}

func pushUp(l, r status) status {
	lSum := max(l.lSum, l.iSum+r.lSum)
	rSum := max(r.rSum, l.rSum+r.iSum)
	iSum := l.iSum + r.iSum
	mSum := max(max(l.mSum, r.mSum), l.rSum + r.lSum)
	return status{lSum, rSum, iSum, mSum}
}

func get(arr []int, l, r int) status {
	if l == r {
		return status{arr[l], arr[l], arr[l], arr[l]}
	}
	m := (l + r) / 2
	left := get(arr, l, m)
	right := get(arr, m+1, r)
	return pushUp(left, right)
}
func maxSubArray2(nums []int) int {
	return get(nums, 0, len(nums)-1).mSum
}
